### PR TITLE
Implement support for input lists in graph_op, deabstraction and graph lowering

### PR DESCRIPTION
**Pull request by lattner**
Implement support for input lists in graph_op, deabstraction and graph lowering

### CODE REVIEW

**Comment by lattner** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 1]
```diff
@@ -323,7 +320,19 @@ static bool isZeroElementArrayAlloc(TupleExtractInst *tei) {
   if (!calleeName.contains("_allocateUninitialized"))
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: contains or equals?

Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: contains or equals?

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 2]
```diff
@@ -344,14 +353,20 @@ static bool decodeArrayElements(SILValue value,
   if (!elementType) return false;
 
   // Handle zero element pattern.
-  if (auto *tei = dyn_cast<TupleExtractInst>(value))
-    if (isZeroElementArrayAlloc(tei)) {
-      if (arrayInsts) {
-        arrayInsts->insert(tei);
-        arrayInsts->insert(tei->getOperand()->getDefiningInstruction());
+  if (auto *tei = dyn_cast<TupleExtractInst>(value)) {
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: do we have a unit test on zero element pattern?

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 3]
```diff
@@ -1614,6 +1646,143 @@ GraphOperationDecoder::decodeAttributeName(Identifier name) {
   return { nameStr.substr(0, dollarLoc), opClass };
 }
 
+/// Analyze the array users of an _allocateUninitialized call, to see if they
+/// are all simple things we can remove.  If so, return add them all to
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: remove "return"

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 4]
```diff
@@ -1614,6 +1646,143 @@ GraphOperationDecoder::decodeAttributeName(Identifier name) {
...
+/// arrayInsts and return false.  If so, return true.
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: if so -> if not?

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 5]
```diff
@@ -1614,6 +1646,143 @@ GraphOperationDecoder::decodeAttributeName(Identifier name) {
...
+/// Given a SILValue that may be an array literal, attempt to decode it into the
+/// values that make up its elements.  If this fails or if the value is not an
+/// array, this returns a null Type.  Otherwise it decodes the array, returns
+/// the values of each element, and returns the element type of the array.
+///
+/// If arrayInsts is non-null and if decoding succeeds, this function adds
+/// all of the instructions relevant to the definition of this array into
+/// the set.  If decoding fails, then the contents of this set is undefined.
+Type GraphOperationInfo::
+decodeArrayElements(SILValue value, SmallVectorImpl<SILValue> &elements,
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: should this function belong to TFConstExpr for more generality? e.g. #assert can then reason about the array elements.

**Review by mhong — APPROVED** [review]
Nice!

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 1 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: contains.  This is a mangled name unfortunately.  I'm not sure what the best way is to model this.

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 5 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Yes, this is on my todo list.  I will move it as I figure out the right place for it.  This is a cut down (and more principled) version of the code used to analyze arrays after the performance optimizer.

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 4 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: fixed, thanks!

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 3 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Fixed!

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 2 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Thanks, I will when I get a chance.

**Pull request merged**